File: backend-v2/frontend-v2/services/calendar-export.ts
/**
 * Calendar Export Service
 * 
 * Frontend service for calendar export and synchronization functionality.
 * Provides a clean interface for interacting with the calendar export APIs.
 */

import { apiRequest } from '../lib/api'

// Types
export interface ExportOptions {
  format: 'ical' | 'csv' | 'json' | 'google_calendar' | 'outlook'
  privacy_level: 'full' | 'business' | 'minimal' | 'anonymous'
  start_date: string
  end_date: string
  barber_ids?: number[]
  service_ids?: number[]
  include_cancelled?: boolean
  include_completed?: boolean
  timezone?: string
  custom_title?: string
  include_client_contact?: boolean
  include_pricing?: boolean
}

export interface ExportResult {
  success: boolean
  format: string
  filename: string
  size_bytes: number
  export_count: number
  export_id: string
  subscription_url?: string
  download_url?: string
  errors: string[]
  warnings: string[]
}

export interface BulkExportOptions {
  barber_ids: number[]
  export_options: ExportOptions
}

export interface BulkExportResult {
  success: boolean
  total_exports: number
  successful_exports: number
  failed_exports: number
  results: ExportResult[]
}

export interface SyncOptions {
  provider: 'google_calendar' | 'outlook' | 'apple_calendar' | 'caldav'
  external_calendar_id: string
  sync_direction?: 'export_only' | 'import_only' | 'bidirectional'
  conflict_resolution?: 'prompt' | 'local_wins' | 'remote_wins' | 'newest_wins' | 'merge'
  sync_frequency?: number
  privacy_level?: 'full' | 'business' | 'minimal' | 'anonymous'
  auto_create_calendar?: boolean
  webhook_enabled?: boolean
}

export interface SyncStatus {
  total_configurations: number
  active_configurations: number
  last_sync_times: Record<string, string | null>
  next_sync_times: Record<string, string | null>
  sync_errors: Record<string, string[]>
  sync_health: Record<string, {
    score: number
    status: string
    last_successful_sync?: string
    error_count: number
    uptime_percentage: number
  }>
  recent_activity: Array<{
    timestamp: string
    action: string
    provider: string
    events_processed: number
    success: boolean
  }>
}

export interface SyncConflict {
  id: string
  local_event: {
    id: string
    title: string
    start_time: string
    end_time: string
    description: string
  }
  remote_event: {
    id: string
    title: string
    start_time: string
    end_time: string
    description: string
  }
  conflict_type: 'time_overlap' | 'content_mismatch' | 'deletion_conflict'
  detected_at: string
  suggested_resolution?: string
}

export interface ConflictData {
  total_conflicts: number
  pending_conflicts: number
  resolved_conflicts: number
  conflicts: SyncConflict[]
}

export interface SubscriptionOptions {
  name: string
  description: string
  privacy_level: 'full' | 'business' | 'minimal' | 'anonymous'
  filters: Record<string, any>
  expires_in_days?: number
}

export interface CalendarSubscription {
  success: boolean
  subscription_id: string
  name: string
  url: string
  expires_at?: string
  created_at: string
}

export interface Analytics {
  export_analytics: {
    total_exports: number
    exports_by_format: Record<string, number>
    exports_by_privacy_level: Record<string, number>
    subscription_usage: Record<string, number>
    sync_performance: Record<string, any>
    most_exported_date_ranges: string[]
    error_rate: number
  }
  sync_analytics: {
    sync_frequency: {
      successful_syncs: number
      failed_syncs: number
      average_duration_seconds: number
      success_rate_percentage: number
    }
    data_volume: {
      events_synced_total: number
      events_created: number
      events_updated: number
      events_deleted: number
    }
    conflicts: {
      total_conflicts: number
      resolved_conflicts: number
      pending_conflicts: number
      conflict_types: Record<string, number>
    }
    provider_breakdown: Record<string, any>
    error_analysis: {
      most_common_errors: string[]
      error_trends: any[]
    }
    performance_metrics: {
      average_sync_time: number
      sync_reliability_score: number
      webhook_response_time: number
    }
  }
  period_days: number
  generated_at: string
}

/**
 * Calendar Export Service Class
 */
export class CalendarExportService {
  private baseUrl = '/api/v2/calendar'

  /**
   * Export calendar in specified format
   */
  async exportCalendar(options: ExportOptions): Promise<ExportResult> {
    try {
      const response = await apiRequest(`${this.baseUrl}/export`, {
        method: 'POST',
        body: JSON.stringify(options),
        headers: {
          'Content-Type': 'application/json'
        }
      })
      return response
    } catch (error) {
      console.error('Calendar export failed:', error)
      throw new Error(
        error.detail || 'Calendar export failed'
      )
    }
  }

  /**
   * Export calendars for multiple barbers
   */
  async bulkExport(options: BulkExportOptions): Promise<BulkExportResult> {
    try {
      const response = await apiRequest(`${this.baseUrl}/bulk-export`, {
        method: 'POST',
        body: JSON.stringify(options),
        headers: {
          'Content-Type': 'application/json'
        }
      })
      return response
    } catch (error) {
      console.error('Bulk export failed:', error)
      throw new Error(
        error.detail || 'Bulk export failed'
      )
    }
  }

  /**
   * Download exported calendar file
   */
  async downloadExport(exportId: string): Promise<Blob> {
    try {
      const response = await apiClient.get(`${this.baseUrl}/download/${exportId}`, {
        responseType: 'blob'
      })
      return response.data
    } catch (error) {
      console.error('Download failed:', error)
      throw new Error(
        error.response?.data?.detail || 'Download failed'
      )
    }
  }

  /**
   * Set up calendar synchronization
   */
  async setupSync(options: SyncOptions): Promise<any> {
    try {
      const response = await apiClient.post(`${this.baseUrl}/sync/setup`, options)
      return response.data
    } catch (error) {
      console.error('Sync setup failed:', error)
      throw new Error(
        error.response?.data?.detail || 'Sync setup failed'
      )
    }
  }

  /**
   * Get sync status
   */
  async getSyncStatus(): Promise<SyncStatus> {
    try {
      const response = await apiClient.get(`${this.baseUrl}/sync/status`)
      return response.data
    } catch (error) {
      console.error('Failed to get sync status:', error)
      throw new Error(
        error.response?.data?.detail || 'Failed to get sync status'
      )
    }
  }

  /**
   * Manually trigger sync
   */
  async triggerSync(configId: string, force = false): Promise<any> {
    try {
      const response = await apiClient.post(
        `${this.baseUrl}/sync/${configId}/trigger`,
        null,
        { params: { force } }
      )
      return response.data
    } catch (error) {
      console.error('Manual sync failed:', error)
      throw new Error(
        error.response?.data?.detail || 'Manual sync failed'
      )
    }
  }

  /**
   * Pause sync
   */
  async pauseSync(configId: string): Promise<any> {
    try {
      const response = await apiClient.post(`${this.baseUrl}/sync/${configId}/pause`)
      return response.data
    } catch (error) {
      console.error('Failed to pause sync:', error)
      throw new Error(
        error.response?.data?.detail || 'Failed to pause sync'
      )
    }
  }

  /**
   * Resume sync
   */
  async resumeSync(configId: string): Promise<any> {
    try {
      const response = await apiClient.post(`${this.baseUrl}/sync/${configId}/resume`)
      return response.data
    } catch (error) {
      console.error('Failed to resume sync:', error)
      throw new Error(
        error.response?.data?.detail || 'Failed to resume sync'
      )
    }
  }

  /**
   * Get sync conflicts
   */
  async getSyncConflicts(): Promise<ConflictData> {
    try {
      const response = await apiClient.get(`${this.baseUrl}/sync/conflicts`)
      return response.data
    } catch (error) {
      console.error('Failed to get conflicts:', error)
      throw new Error(
        error.response?.data?.detail || 'Failed to get conflicts'
      )
    }
  }

  /**
   * Resolve sync conflicts
   */
  async resolveConflicts(
    conflictIds: string[],
    resolutionStrategy: 'prompt' | 'local_wins' | 'remote_wins' | 'newest_wins' | 'merge',
    userChoices?: Record<string, any>
  ): Promise<any> {
    try {
      const response = await apiClient.post(`${this.baseUrl}/sync/conflicts/resolve`, {
        conflict_ids: conflictIds,
        resolution_strategy: resolutionStrategy,
        user_choices: userChoices
      })
      return response.data
    } catch (error) {
      console.error('Conflict resolution failed:', error)
      throw new Error(
        error.response?.data?.detail || 'Conflict resolution failed'
      )
    }
  }

  /**
   * Create calendar subscription
   */
  async createSubscription(options: SubscriptionOptions): Promise<CalendarSubscription> {
    try {
      const response = await apiClient.post(`${this.baseUrl}/subscription`, options)
      return response.data
    } catch (error) {
      console.error('Subscription creation failed:', error)
      throw new Error(
        error.response?.data?.detail || 'Subscription creation failed'
      )
    }
  }

  /**
   * Get analytics
   */
  async getAnalytics(days = 30): Promise<Analytics> {
    try {
      const response = await apiClient.get(`${this.baseUrl}/analytics`, {
        params: { days }
      })
      return response.data
    } catch (error) {
      console.error('Failed to get analytics:', error)
      throw new Error(
        error.response?.data?.detail || 'Failed to get analytics'
      )
    }
  }

  /**
   * Helper method to download file with proper filename
   */
  async downloadFile(exportId: string, filename?: string): Promise<void> {
    try {
      const blob = await this.downloadExport(exportId)
      
      // Create download link
      const url = window.URL.createObjectURL(blob)
      const link = document.createElement('a')
      link.href = url
      link.download = filename || `calendar-export-${exportId}`
      
      // Trigger download
      document.body.appendChild(link)
      link.click()
      document.body.removeChild(link)
      
      // Clean up
      window.URL.revokeObjectURL(url)
    } catch (error) {
      console.error('File download failed:', error)
      throw error
    }
  }

  /**
   * Get export format display name
   */
  getFormatDisplayName(format: string): string {
    const formatNames: Record<string, string> = {
      ical: 'iCalendar (.ics)',
      csv: 'CSV Spreadsheet (.csv)',
      json: 'JSON Data (.json)',
      google_calendar: 'Google Calendar',
      outlook: 'Outlook Calendar (.msg)'
    }
    return formatNames[format] || format
  }

  /**
   * Get privacy level description
   */
  getPrivacyLevelDescription(level: string): string {
    const descriptions: Record<string, string> = {
      full: 'All details including client contact information',
      business: 'Business details without sensitive client info',
      minimal: 'Just appointment times and services',
      anonymous: 'Generic placeholder appointments'
    }
    return descriptions[level] || level
  }

  /**
   * Validate export options
   */
  validateExportOptions(options: ExportOptions): string[] {
    const errors: string[] = []

    // Check required fields
    if (!options.format) {
      errors.push('Export format is required')
    }

    if (!options.start_date) {
      errors.push('Start date is required')
    }

    if (!options.end_date) {
      errors.push('End date is required')
    }

    // Check date range
    if (options.start_date && options.end_date) {
      const start = new Date(options.start_date)
      const end = new Date(options.end_date)
      
      if (start >= end) {
        errors.push('End date must be after start date')
      }

      // Check for reasonable date range (max 1 year)
      const yearInMs = 365 * 24 * 60 * 60 * 1000
      if (end.getTime() - start.getTime() > yearInMs) {
        errors.push('Date range cannot exceed 1 year')
      }
    }

    // Check barber IDs if provided
    if (options.barber_ids && options.barber_ids.length === 0) {
      errors.push('At least one barber must be selected if barber filter is used')
    }

    return errors
  }

  /**
   * Format sync frequency for display
   */
  formatSyncFrequency(minutes: number): string {
    if (minutes < 60) {
      return `${minutes} minutes`
    } else if (minutes < 1440) {
      const hours = Math.floor(minutes / 60)
      return `${hours} hour${hours > 1 ? 's' : ''}`
    } else {
      const days = Math.floor(minutes / 1440)
      return `${days} day${days > 1 ? 's' : ''}`
    }
  }

  /**
   * Get sync health status color
   */
  getSyncHealthColor(score: number): string {
    if (score >= 90) return 'green'
    if (score >= 70) return 'yellow'
    if (score >= 50) return 'orange'
    return 'red'
  }

  /**
   * Format file size for display
   */
  formatFileSize(bytes: number): string {
    if (bytes === 0) return '0 Bytes'
    
    const k = 1024
    const sizes = ['Bytes', 'KB', 'MB', 'GB']
    const i = Math.floor(Math.log(bytes) / Math.log(k))
    
    return parseFloat((bytes / Math.pow(k, i)).toFixed(2)) + ' ' + sizes[i]
  }
}

// Export singleton instance
export const calendarExportService = new CalendarExportService()

File: backend-v2/routers/calendar_export.py
"""
Calendar Export and Sync API Endpoints

Comprehensive API endpoints for calendar export and synchronization features:
- Multiple export formats (iCal, CSV, JSON, Google Calendar, Outlook)
- Calendar subscription management
- Two-way synchronization setup and management
- Conflict resolution
- Analytics and monitoring
"""

import io
import json
import logging
from datetime import datetime, timedelta
from typing import List, Dict, Any, Optional, Tuple
from fastapi import APIRouter, Depends, HTTPException, status, Response, BackgroundTasks
from fastapi.responses import StreamingResponse, PlainTextResponse
from sqlalchemy.orm import Session
from pydantic import BaseModel, Field

from dependencies import get_db, get_current_user
from models import User
from services.calendar_export_service import (
    CalendarExportService, 
    ExportFormat, 
    PrivacyLevel, 
    ExportOptions,
    SyncProvider,
    SyncOptions
)
from services.calendar_sync_service import (
    CalendarSyncService,
    SyncDirection,
    ConflictResolution
)

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/v2/calendar", tags=["calendar-export"])


# Request Models
class CalendarExportRequest(BaseModel):
    """Request model for calendar export."""
    format: str = Field(..., description="Export format: ical, csv, json, google_calendar, outlook")
    privacy_level: str = Field(default="business", description="Privacy level: full, business, minimal, anonymous")
    start_date: datetime = Field(..., description="Start date for export")
    end_date: datetime = Field(..., description="End date for export")
    barber_ids: Optional[List[int]] = Field(None, description="Specific barber IDs to include")
    service_ids: Optional[List[int]] = Field(None, description="Specific service IDs to include")
    include_cancelled: bool = Field(default=False, description="Include cancelled appointments")
    include_completed: bool = Field(default=True, description="Include completed appointments")
    timezone: str = Field(default="UTC", description="Timezone for export")
    custom_title: Optional[str] = Field(None, description="Custom calendar title")
    include_client_contact: bool = Field(default=False, description="Include client contact info")
    include_pricing: bool = Field(default=False, description="Include pricing information")


class BulkExportRequest(BaseModel):
    """Request model for bulk export of multiple barbers."""
    barber_ids: List[int] = Field(..., description="List of barber IDs to export")
    export_options: CalendarExportRequest = Field(..., description="Export options to apply")


class SyncSetupRequest(BaseModel):
    """Request model for setting up calendar synchronization."""
    provider: str = Field(..., description="Sync provider: google_calendar, outlook, apple_calendar, caldav")
    external_calendar_id: str = Field(..., description="External calendar ID")
    sync_direction: str = Field(default="bidirectional", description="Sync direction: export_only, import_only, bidirectional")
    conflict_resolution: str = Field(default="prompt", description="Conflict resolution: prompt, local_wins, remote_wins, newest_wins, merge")
    sync_frequency: int = Field(default=15, description="Sync frequency in minutes")
    privacy_level: str = Field(default="business", description="Privacy level for sync")
    auto_create_calendar: bool = Field(default=True, description="Auto-create calendar if not exists")
    webhook_enabled: bool = Field(default=True, description="Enable webhook notifications")


class ConflictResolutionRequest(BaseModel):
    """Request model for conflict resolution."""
    conflict_ids: List[str] = Field(..., description="List of conflict IDs to resolve")
    resolution_strategy: str = Field(..., description="Resolution strategy")
    user_choices: Optional[Dict[str, Any]] = Field(None, description="User choices for manual resolution")


class SubscriptionCreateRequest(BaseModel):
    """Request model for creating calendar subscription."""
    name: str = Field(..., description="Subscription name")
    description: str = Field(..., description="Subscription description")
    privacy_level: str = Field(default="business", description="Privacy level")
    filters: Dict[str, Any] = Field(default_factory=dict, description="Calendar filters")
    expires_in_days: Optional[int] = Field(None, description="Expiration in days")


# Response Models
class ExportResponse(BaseModel):
    """Response model for calendar export."""
    success: bool
    format: str
    filename: str
    size_bytes: int
    export_count: int
    export_id: str
    subscription_url: Optional[str] = None
    download_url: Optional[str] = None
    errors: List[str] = []
    warnings: List[str] = []


class BulkExportResponse(BaseModel):
    """Response model for bulk export."""
    success: bool
    total_exports: int
    successful_exports: int
    failed_exports: int
    results: List[ExportResponse]


class SyncStatusResponse(BaseModel):
    """Response model for sync status."""
    total_configurations: int
    active_configurations: int
    last_sync_times: Dict[str, Optional[str]]
    next_sync_times: Dict[str, Optional[str]]
    sync_errors: Dict[str, List[str]]
    sync_health: Dict[str, Dict[str, Any]]
    recent_activity: List[Dict[str, Any]]


class ConflictResponse(BaseModel):
    """Response model for sync conflicts."""
    total_conflicts: int
    pending_conflicts: int
    resolved_conflicts: int
    conflicts: List[Dict[str, Any]]


# API Endpoints

@router.post("/export", response_model=ExportResponse)
async def export_calendar(
    request: CalendarExportRequest,
    background_tasks: BackgroundTasks,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Export calendar in specified format with privacy controls."""
    try:
        # Validate export format
        try:
            export_format = ExportFormat(request.format.lower())
        except ValueError:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Unsupported export format: {request.format}"
            )
        
        # Validate privacy level
        try:
            privacy_level = PrivacyLevel(request.privacy_level.lower())
        except ValueError:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Invalid privacy level: {request.privacy_level}"
            )
        
        # Create export options
        export_options = ExportOptions(
            format=export_format,
            privacy_level=privacy_level,
            start_date=request.start_date,
            end_date=request.end_date,
            barber_ids=request.barber_ids,
            service_ids=request.service_ids,
            include_cancelled=request.include_cancelled,
            include_completed=request.include_completed,
            timezone=request.timezone,
            custom_title=request.custom_title,
            include_client_contact=request.include_client_contact,
            include_pricing=request.include_pricing
        )
        
        # Perform export
        export_service = CalendarExportService(db)
        result = export_service.export_calendar(current_user, export_options)
        
        if not result.success:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail=f"Export failed: {'; '.join(result.errors)}"
            )
        
        # Store export data temporarily for download
        download_url = f"/api/v2/calendar/download/{result.export_id}"
        background_tasks.add_task(_store_export_data, result.export_id, result.content, result.filename)
        
        return ExportResponse(
            success=result.success,
            format=result.format.value,
            filename=result.filename,
            size_bytes=result.size_bytes,
            export_count=result.export_count,
            export_id=result.export_id,
            subscription_url=result.subscription_url,
            download_url=download_url,
            errors=result.errors or [],
            warnings=result.warnings or []
        )
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Calendar export failed for user {current_user.id}: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Export failed: {str(e)}"
        )


@router.get("/download/{export_id}")
async def download_export(
    export_id: str,
    current_user: User = Depends(get_current_user)
):
    """Download exported calendar file."""
    try:
        # Retrieve stored export data
        export_data = _get_export_data(export_id)
        
        if not export_data:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Export not found or expired"
            )
        
        content, filename, content_type = export_data
        
        # Determine content type based on file extension
        if filename.endswith('.ics'):
            media_type = "text/calendar"
        elif filename.endswith('.csv'):
            media_type = "text/csv"
        elif filename.endswith('.json'):
            media_type = "application/json"
        else:
            media_type = "application/octet-stream"
        
        # Create streaming response
        return StreamingResponse(
            io.BytesIO(content.encode() if isinstance(content, str) else content),
            media_type=media_type,
            headers={"Content-Disposition": f"attachment; filename={filename}"}
        )
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Download failed for export {export_id}: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Download failed"
        )


@router.post("/bulk-export", response_model=BulkExportResponse)
async def bulk_export_multiple_barbers(
    request: BulkExportRequest,
    background_tasks: BackgroundTasks,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Export calendars for multiple barbers."""
    try:
        # Validate export format and privacy level
        export_format = ExportFormat(request.export_options.format.lower())
        privacy_level = PrivacyLevel(request.export_options.privacy_level.lower())
        
        # Create base export options
        base_options = ExportOptions(
            format=export_format,
            privacy_level=privacy_level,
            start_date=request.export_options.start_date,
            end_date=request.export_options.end_date,
            service_ids=request.export_options.service_ids,
            include_cancelled=request.export_options.include_cancelled,
            include_completed=request.export_options.include_completed,
            timezone=request.export_options.timezone,
            custom_title=request.export_options.custom_title,
            include_client_contact=request.export_options.include_client_contact,
            include_pricing=request.export_options.include_pricing
        )
        
        # Perform bulk export
        export_service = CalendarExportService(db)
        results = export_service.bulk_export_multiple_barbers(
            current_user, 
            request.barber_ids, 
            base_options
        )
        
        # Convert results to response format
        export_responses = []
        successful_exports = 0
        
        for result in results:
            if result.success:
                successful_exports += 1
                download_url = f"/api/v2/calendar/download/{result.export_id}"
                background_tasks.add_task(_store_export_data, result.export_id, result.content, result.filename)
            else:
                download_url = None
            
            export_responses.append(ExportResponse(
                success=result.success,
                format=result.format.value,
                filename=result.filename,
                size_bytes=result.size_bytes,
                export_count=result.export_count,
                export_id=result.export_id,
                subscription_url=result.subscription_url,
                download_url=download_url,
                errors=result.errors or [],
                warnings=result.warnings or []
            ))
        
        return BulkExportResponse(
            success=successful_exports > 0,
            total_exports=len(results),
            successful_exports=successful_exports,
            failed_exports=len(results) - successful_exports,
            results=export_responses
        )
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Bulk export failed for user {current_user.id}: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Bulk export failed: {str(e)}"
        )


@router.post("/sync/setup")
async def setup_calendar_sync(
    request: SyncSetupRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Set up calendar synchronization with external provider."""
    try:
        # Validate provider
        try:
            provider = SyncProvider(request.provider.lower())
        except ValueError:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Unsupported sync provider: {request.provider}"
            )
        
        # Create sync options
        sync_options = SyncOptions(
            provider=provider,
            calendar_id=request.external_calendar_id,
            sync_direction=request.sync_direction,
            conflict_resolution=request.conflict_resolution,
            sync_frequency=request.sync_frequency,
            privacy_level=PrivacyLevel(request.privacy_level.lower()),
            auto_create_calendar=request.auto_create_calendar,
            webhook_enabled=request.webhook_enabled
        )
        
        # Set up sync
        sync_service = CalendarSyncService(db)
        result = await sync_service.setup_sync_configuration(
            current_user,
            provider,
            request.external_calendar_id,
            sync_options
        )
        
        if not result.get("success"):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=result.get("error", "Sync setup failed")
            )
        
        return result
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Sync setup failed for user {current_user.id}: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Sync setup failed: {str(e)}"
        )


@router.get("/sync/status", response_model=SyncStatusResponse)
async def get_sync_status(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Get comprehensive sync status for user."""
    try:
        sync_service = CalendarSyncService(db)
        status_data = await sync_service.get_sync_status(current_user)
        
        return SyncStatusResponse(**status_data)
        
    except Exception as e:
        logger.error(f"Failed to get sync status for user {current_user.id}: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to get sync status: {str(e)}"
        )


@router.post("/sync/{config_id}/trigger")
async def trigger_sync(
    config_id: str,
    force: bool = False,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Manually trigger calendar synchronization."""
    try:
        sync_service = CalendarSyncService(db)
        result = await sync_service.perform_sync(config_id, force=force)
        
        return {
            "success": result.success,
            "sync_id": result.sync_id,
            "events_processed": result.events_processed,
            "events_created": result.events_created,
            "events_updated": result.events_updated,
            "conflicts_detected": result.conflicts_detected,
            "conflicts_resolved": result.conflicts_resolved,
            "duration_seconds": (result.completed_at - result.started_at).total_seconds(),
            "errors": result.errors,
            "warnings": result.warnings,
            "next_sync_at": result.next_sync_at.isoformat() if result.next_sync_at else None
        }
        
    except Exception as e:
        logger.error(f"Manual sync failed for config {config_id}: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Sync failed: {str(e)}"
        )


@router.post("/sync/{config_id}/pause")
async def pause_sync(
    config_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Pause calendar synchronization."""
    try:
        sync_service = CalendarSyncService(db)
        result = await sync_service.pause_sync(config_id)
        
        if not result.get("success"):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=result.get("error", "Failed to pause sync")
            )
        
        return result
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Failed to pause sync {config_id}: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to pause sync: {str(e)}"
        )


@router.post("/sync/{config_id}/resume")
async def resume_sync(
    config_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Resume calendar synchronization."""
    try:
        sync_service = CalendarSyncService(db)
        result = await sync_service.resume_sync(config_id)
        
        if not result.get("success"):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=result.get("error", "Failed to resume sync")
            )
        
        return result
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Failed to resume sync {config_id}: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to resume sync: {str(e)}"
        )


@router.get("/sync/conflicts", response_model=ConflictResponse)
async def get_sync_conflicts(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Get pending sync conflicts for user."""
    try:
        # This would query conflicts from database
        conflicts_data = {
            "total_conflicts": 0,
            "pending_conflicts": 0,
            "resolved_conflicts": 0,
            "conflicts": []
        }
        
        return ConflictResponse(**conflicts_data)
        
    except Exception as e:
        logger.error(f"Failed to get conflicts for user {current_user.id}: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to get conflicts: {str(e)}"
        )


@router.post("/sync/conflicts/resolve")
async def resolve_sync_conflicts(
    request: ConflictResolutionRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Resolve sync conflicts."""
    try:
        sync_service = CalendarSyncService(db)
        
        # This would resolve conflicts based on the request
        results = {
            "resolved_count": 0,
            "pending_count": 0,
            "error_count": 0,
            "resolutions": []
        }
        
        return results
        
    except Exception as e:
        logger.error(f"Conflict resolution failed for user {current_user.id}: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Conflict resolution failed: {str(e)}"
        )


@router.post("/subscription", response_model=Dict[str, Any])
async def create_calendar_subscription(
    request: SubscriptionCreateRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Create a shareable calendar subscription."""
    try:
        privacy_level = PrivacyLevel(request.privacy_level.lower())
        
        export_service = CalendarExportService(db)
        subscription = export_service.create_subscription_calendar(
            current_user,
            request.name,
            request.description,
            privacy_level,
            request.filters,
            request.expires_in_days
        )
        
        return {
            "success": True,
            "subscription_id": subscription.id,
            "name": subscription.name,
            "url": subscription.url,
            "expires_at": subscription.expires_at.isoformat() if subscription.expires_at else None,
            "created_at": subscription.created_at.isoformat()
        }
        
    except Exception as e:
        logger.error(f"Subscription creation failed for user {current_user.id}: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Subscription creation failed: {str(e)}"
        )


@router.get("/subscription/{subscription_id}.ics")
async def get_subscription_calendar(
    subscription_id: str,
    response: Response
):
    """Get calendar data for subscription URL."""
    try:
        export_service = CalendarExportService(None)  # No DB session needed for subscription
        calendar_data = export_service.get_subscription_calendar(subscription_id)
        
        response.headers["Content-Type"] = "text/calendar; charset=utf-8"
        response.headers["Content-Disposition"] = f"attachment; filename=calendar-{subscription_id}.ics"
        
        return PlainTextResponse(calendar_data, media_type="text/calendar")
        
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(e)
        )
    except Exception as e:
        logger.error(f"Subscription calendar failed for {subscription_id}: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to generate subscription calendar"
        )


@router.get("/analytics")
async def get_export_analytics(
    days: int = 30,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Get calendar export and sync analytics."""
    try:
        export_service = CalendarExportService(db)
        sync_service = CalendarSyncService(db)
        
        export_analytics = export_service.get_export_analytics(current_user, days)
        sync_analytics = await sync_service.get_sync_analytics(current_user, days)
        
        return {
            "export_analytics": export_analytics,
            "sync_analytics": sync_analytics,
            "period_days": days,
            "generated_at": datetime.utcnow().isoformat()
        }
        
    except Exception as e:
        logger.error(f"Analytics failed for user {current_user.id}: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Analytics failed: {str(e)}"
        )


@router.post("/webhooks/{provider}/{user_id}")
async def handle_calendar_webhook(
    provider: str,
    user_id: int,
    webhook_data: Dict[str, Any],
    db: Session = Depends(get_db)
):
    """Handle webhook notifications from external calendar providers."""
    try:
        sync_provider = SyncProvider(provider.lower())
        
        sync_service = CalendarSyncService(db)
        result = await sync_service.handle_webhook(sync_provider, user_id, webhook_data)
        
        return result
        
    except ValueError:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Unsupported provider: {provider}"
        )
    except Exception as e:
        logger.error(f"Webhook handling failed for {provider}/{user_id}: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Webhook processing failed"
        )


# Helper functions

def _store_export_data(export_id: str, content: str, filename: str):
    """Store export data temporarily for download."""
    # This would store in Redis or temporary file system
    # For now, we'll just log it
    logger.info(f"Stored export data for {export_id}: {filename}")


def _get_export_data(export_id: str) -> Optional[Tuple[str, str, str]]:
    """Retrieve stored export data."""
    # This would retrieve from Redis or temporary file system
    # For now, return None (not found)
    return None